### ISSUE TITLE
typechecker:handle_type_error function_clause again

### ISSUE THREAD

**Issue by erszcz** [josefs/Gradualizer#148]
I've just found this:

```sh
$ gradualizer src/history_archiver.erl
...
src/history_archiver.erl: escript: exception error: no function clause matching
                 typechecker:handle_type_error({type_error,
                                                {map,201,
                                                 {var,201,'Ctx'},
                                                 [{map_field_exact,201,
                                                   {atom,201,archive_id},
                                                   {var,201,'ID'}}]},
                                                {type,153,map,
                                                 [{type,153,map_field_exact,
                                                   [{atom,153,'__struct__'},
                                                    {atom,153,mam_context}]},
                                                  {type,154,map_field_exact,
                                                   [{atom,154,from},
                                                    {type,154,any,[]}]},
                                                  {type,155,map_field_exact,
                                                   [{atom,155,to},
                                                    {type,155,any,[]}]},
                                                  {type,156,map_field_exact,
                                                   [{atom,156,message},
                                                    {remote_type,156,
                                                     [{atom,156,exml},
                                                      {atom,156,element},
                                                      []]}]},
                                                  {type,157,map_field_exact,
                                                   [{atom,157,archive_id},
                                                    {type,157,union,
                                                     [{atom,157,undefined},
                                                      {type,157,binary,
                                                       []}]}]}]},
                                                {user_type,0,
                                                 mam_context_with_id,[]}}) (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 3734)
  in function  typechecker:'-type_check_forms/2-fun-0-'/7 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 3613)
  in call from lists:foldr/3 (lists.erl, line 1276)
  in call from gradualizer:type_check_forms/3 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/gradualizer.erl, line 147)
  in call from gradualizer:'-type_check_files/2-fun-1-'/4 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/gradualizer.erl, line 118)
  in call from lists:foldl/3 (lists.erl, line 1263)
  in call from gradualizer_cli:handle_args/1 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/gradualizer_cli.erl, line 17)
```

It seems similar to #136 and might be caused by the same problem with measuring coverage on `throw`s. I'll try to come up with a minimal test case when I have a bit more time.

**Comment by zuiderkwast** [comment]
It looks like it's a map update where something else is expected.
```Erlang
-spec f(map()) -> integer().
f(M) -> M#{piff => paff}.
```

**Comment by gomoripeti** [comment]
What do you think if the formatted error would be changed from `The map #{piff => paff} is expected...` to `The expression #{piff => paff} is expected...` (for all expression types). This way matching on the expression to fetch it's type (basically the first element of the tuple) is unnecessary, and we could further merge some `handle_type_error` clauses?

**Comment by zuiderkwast** [comment]
Sure! (The expression is `M#{piff => paff}` though, a map update expression.)

But I have another idea for the formatting helper function:

```Erlang
describe_expr({map, _, _}) -> "map";
describe_expr({map, _, _, _}) -> "map update";
describe_expr({var, _, _}) -> "variable";
% ...
describe_expr(_) -> "expression".
```

Then we can also include the pretty-printed expression as a whole if it is less than e.g. 20 characters.

**Comment by Zalastax** [comment]
The describe_expr is a good idea in the long run but it doesn't help us at the moment. Let's wait until all expressions are thrown in the same way and can be distinguished easily in a single pattern. I don't want to duplicate the pattern that checks if the thrown error regards e.g. a map update.

**Comment by zuiderkwast** [comment]
I'm closing this as it (the crash because of missing clause in `handle_type_error/1`) is fixed by #151.

The new problem discovered is a different problem and a known problem test case has been added for that already.
